File: deno-knowledge-app/src/collectors/knowledge.ts
// collect_knowledge — mirrors knowledge.py collect_knowledge
// Inspects knowledge/: decisions, lektion-*, memory/*, vault CHANGELOG tail.

import { readText } from "../shared.ts";
import { join } from "@std/path";
import { repoRoot } from "./project.ts";

interface Decision {
  id: string;
  title: string;
  status: string;
  body: string;
  mtime?: number; // epoch seconds
}

export interface MemoryEntry {
  name: string;
  mtime: number; // epoch seconds
}

export interface LektionEntry {
  name: string;
  mtime: number; // epoch seconds
}

const DECISION_LINE_RE = /^\s*##\s+(?<id>\d+)\s*(?:—|-{1,2})\s*(?<title>.*\S)\s*$/;
const STATUS_RE = /^\s*(?:[-*]\s*)?\**\s*Status\s*\**\s*[:=]\s*(.+\S)\s*$/i;

export function parseDecisionsMd(text: string): Decision[] {
  const out: Decision[] = [];
  let current: Decision | null = null;
  const body: string[] = [];

  function flush() {
    if (current !== null) {
      current.body = body.join("\n").trim();
    }
  }

  for (const line of text.split("\n")) {
    const m = line.match(DECISION_LINE_RE);
    if (m) {
      flush();
      body.length = 0;
      current = { id: m.groups!.id, title: m.groups!.title, status: "", body: "" };
      out.push(current);
      continue;
    }
    if (current !== null) {
      const sm = line.match(STATUS_RE);
      if (sm && !current.status) {
        current.status = sm[1].trim().replace(/^\*+|\*+$/g, "").trim();
      }
      body.push(line);
    }
  }
  flush();
  return out;
}

async function backlogDecisions(repo: string): Promise<Decision[]> {
  const ddir = join(repo, "backlog", "decisions");
  try {
    const out: Decision[] = [];
    for await (const e of Deno.readDir(ddir)) {
      if (!e.isFile || !e.name.match(/^decision-\d+/)) continue;
      const fpath = join(ddir, e.name);
      const m = e.name.match(/^decision-(\d+)/);
      const did = m ? m[1] : e.name.replace(/\.md$/, "");
      const text = await readText(fpath) ?? "";
      let title = e.name.replace(/\.md$/, "");
      let status = "";
      for (const line of text.split("\n")) {
        const hm = line.match(/^#\s+(.*\S)\s*$/);
        if (hm) {
          title = hm[1].trim();
          break;
        }
      }
      const sm = text.match(/^status:\s*(.+\S)\s*$/im);
      if (sm) status = sm[1].trim();
      let body = text;
      if (text.trimStart().startsWith("---")) {
        const parts = text.split("---");
        if (parts.length >= 3) body = parts.slice(2).join("---");
      }
      let mtime = 0;
      try {
        const stat = await Deno.stat(fpath);
        mtime = stat.mtime ? Math.floor(stat.mtime.getTime() / 1000) : 0;
      } catch { /* ok */ }
      out.push({ id: did, title, status, body: body.trim().slice(0, 8000), mtime });
    }
    // Sort: mtime DESC (newest first), then id as tiebreaker
    out.sort((a, b) => {
      const md = (b.mtime ?? 0) - (a.mtime ?? 0);
      if (md !== 0) return md;
      return a.id.localeCompare(b.id);
    });
    return out;
  } catch {
    return [];
  }
}

function resolveVault(): string | null {
  const env = Deno.env.get("OBSIDIAN_VAULT_PATH") ?? Deno.env.get("TASKNOTES_VAULT");
  if (env) return env;
  const def = `${Deno.env.get("HOME")}/GITHUB/ObsidianPKM`;
  try {
    Deno.statSync(def);
    return def;
  } catch {
    return null;
  }
}

export interface ChangelogEntry {
  heading: string; // e.g. "## 2024-06-01 — feat: foo"
  body: string; // lines below the heading until next heading
}

/** Parse a CHANGELOG.md into blocks (one block = one `##`/`###` heading + its body).
 *  H1 (`#`) is treated as document title and skipped.
 *  Blocks with empty body are skipped (avoids the document-title artefact).
 *  Result is sorted newest-first: if headings contain ISO dates and the file
 *  appears to be oldest-first (first date < last date), the array is reversed.
 */
export function parseChangelogBlocks(text: string, limit = 20): ChangelogEntry[] {
  const blocks: ChangelogEntry[] = [];
  let current: ChangelogEntry | null = null;
  const bodyLines: string[] = [];

  function flush() {
    if (current !== null) {
      const body = bodyLines.join("\n").trim();
      if (body) {
        current.body = body;
        blocks.push(current);
      }
      current = null;
    }
    bodyLines.length = 0;
  }

  for (const line of text.split("\n")) {
    // Only ## and ### are release-entry headings; # is document title → skip
    if (/^#{2,3}\s+/.test(line)) {
      flush();
      current = { heading: line.trim(), body: "" };
    } else if (/^#\s+/.test(line)) {
      // H1: flush any open block but don't start a new entry
      flush();
    } else if (current !== null) {
      bodyLines.push(line);
    }
  }
  flush();

  // Ensure newest-first: detect oldest-first by comparing ISO-date in first vs last heading.
  const dateRe = /(\d{4}-\d{2}-\d{2})/;
  if (blocks.length >= 2) {
    const firstDate = (blocks[0].heading.match(dateRe) ?? [])[1];
    const lastDate = (blocks[blocks.length - 1].heading.match(dateRe) ?? [])[1];
    if (firstDate && lastDate && firstDate < lastDate) {
      blocks.reverse(); // file was oldest-first → flip to newest-first
    }
  }

  return blocks.slice(0, limit);
}

async function vaultChangelog(
  repo: string,
  repoName: string,
  vault: string | null,
  limit = 20,
): Promise<ChangelogEntry[]> {
  // Prefer repo-local knowledge/CHANGELOG.md first.
  const localPath = join(repo, "knowledge", "CHANGELOG.md");
  const localText = await readText(localPath);
  if (localText) return parseChangelogBlocks(localText, limit);

  // Fall back to vault path (dgk-specific).
  if (!vault) return [];
  const vaultPath = join(vault, "Efforts", "Work", "dgk", repoName, "CHANGELOG.md");
  const vaultText = await readText(vaultPath);
  if (!vaultText) return [];
  return parseChangelogBlocks(vaultText, limit);
}

export interface DocEntry {
  id: string;
  title: string;
  file: string; // filename within backlog/docs/
  mtime?: number; // epoch seconds
}

/** Read backlog/docs/*.md — like backlogDecisions but for docs. */
async function backlogDocs(repo: string): Promise<DocEntry[]> {
  const docsDir = join(repo, "backlog", "docs");
  const out: DocEntry[] = [];
  try {
    for await (const e of Deno.readDir(docsDir)) {
      if (!e.isFile || !e.name.endsWith(".md")) continue;
      const fpath = join(docsDir, e.name);
      const text = await readText(fpath) ?? "";
      // Extract title from frontmatter or first heading
      let title = e.name.replace(/\.md$/, "");
      const fm = text.match(/^---\n([\s\S]*?)\n---/);
      if (fm) {
        const tm = fm[1].match(/^title:\s*(.+\S)\s*$/im);
        if (tm) title = tm[1].replace(/^["']|["']$/g, "").trim();
      }
      if (title === e.name.replace(/\.md$/, "")) {
        for (const line of text.split("\n")) {
          const hm = line.match(/^#\s+(.*\S)\s*$/);
          if (hm) {
            title = hm[1].trim();
            break;
          }
        }
      }
      // id from frontmatter or filename
      let id = e.name.replace(/\.md$/, "");
      if (fm) {
        const im = fm[1].match(/^id:\s*(.+\S)\s*$/im);
        if (im) id = im[1].trim();
      }
      let mtime = 0;
      try {
        const stat = await Deno.stat(fpath);
        mtime = stat.mtime ? Math.floor(stat.mtime.getTime() / 1000) : 0;
      } catch { /* ok */ }
      out.push({ id, title, file: e.name, mtime });
    }
    // Sort: mtime DESC (newest first)
    out.sort((a, b) => {
      const md = (b.mtime ?? 0) - (a.mtime ?? 0);
      if (md !== 0) return md;
      return a.id.localeCompare(b.id);
    });
  } catch { /* no docs dir */ }
  return out;
}

export async function collectKnowledge(
  cwd: string,
  vaultPath?: string,
): Promise<Record<string, unknown>> {
  try {
    const repo = await repoRoot(cwd);
    const kdir = join(repo, "knowledge");

    // Decisions: prefer knowledge/decisions.md, else backlog/decisions/*.
    let decisions: Decision[] = [];
    const decMdPath = join(kdir, "decisions.md");
    const decMdText = await readText(decMdPath);
    if (decMdText) {
      decisions = parseDecisionsMd(decMdText);
      // All decisions from decisions.md share the file mtime
      let decMdMtime = 0;
      try {
        const stat = await Deno.stat(decMdPath);
        decMdMtime = stat.mtime ? Math.floor(stat.mtime.getTime() / 1000) : 0;
      } catch { /* ok */ }
      for (const d of decisions) {
        d.mtime = decMdMtime;
      }
      // Sort: mtime DESC (all same mtime here), then id as tiebreaker
      decisions.sort((a, b) => {
        const md = (b.mtime ?? 0) - (a.mtime ?? 0);
        if (md !== 0) return md;
        return a.id.localeCompare(b.id);
      });
    }
    if (!decisions.length) {
      decisions = await backlogDecisions(repo);
    }

    // lektion-*.md + optional lessons-learned.md — with mtime, sorted newest-first
    const lektionen: LektionEntry[] = [];
    try {
      for await (const e of Deno.readDir(kdir)) {
        if (e.isFile && e.name.match(/^lektion-.*\.md$/)) {
          let mtime = 0;
          try {
            const stat = await Deno.stat(join(kdir, e.name));
            mtime = stat.mtime ? Math.floor(stat.mtime.getTime() / 1000) : 0;
          } catch { /* ok */ }
          lektionen.push({ name: e.name, mtime });
        }
      }
    } catch { /* no kdir */ }
    // Include lessons-learned.md as a special entry if present
    const lessonsFile = join(kdir, "lessons-learned.md");
    try {
      const stat = await Deno.stat(lessonsFile);
      const mtime = stat.mtime ? Math.floor(stat.mtime.getTime() / 1000) : 0;
      lektionen.push({ name: "lessons-learned.md", mtime });
    } catch { /* not present → skip */ }
    // Sort: newest first
    lektionen.sort((a, b) => b.mtime - a.mtime);

    // knowledge/memory/*.md — with mtime, sorted newest-first
    const memory: MemoryEntry[] = [];
    const memDir = join(kdir, "memory");
    try {
      for await (const e of Deno.readDir(memDir)) {
        if (e.isFile && e.name.endsWith(".md")) {
          let mtime = 0;
          try {
            const stat = await Deno.stat(join(memDir, e.name));
            mtime = stat.mtime ? Math.floor(stat.mtime.getTime() / 1000) : 0;
          } catch { /* ok */ }
          memory.push({ name: e.name, mtime });
        }
      }
    } catch { /* no memory dir */ }
    // Sort: newest first
    memory.sort((a, b) => b.mtime - a.mtime);

    // CHANGELOG: prefer knowledge/CHANGELOG.md, fall back to vault path
    const vault = vaultPath ?? resolveVault();
    const repoName = repo.split("/").pop()!;
    const changelog = await vaultChangelog(repo, repoName, vault);

    // backlog/docs/*.md
    const docs = await backlogDocs(repo);

    return { available: true, decisions, lektionen, memory, changelog, docs };
  } catch (exc) {
    return { available: false, reason: `collect_knowledge failed: ${exc}` };
  }
}

File: go-knowledge-app/internal/collectors/knowledge.go
package collectors

import (
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"go-knowledge-app/internal/shared"
)

// Decision mirrors the TS Decision interface.
type Decision struct {
	ID     string `json:"id"`
	Title  string `json:"title"`
	Status string `json:"status"`
	Body   string `json:"body"`
	Mtime  *int   `json:"mtime,omitempty"` // epoch seconds
}

// MemoryEntry / LektionEntry — a knowledge file with its mtime (epoch seconds).
type MemoryEntry struct {
	Name  string `json:"name"`
	Mtime int    `json:"mtime"`
}
type LektionEntry struct {
	Name  string `json:"name"`
	Mtime int    `json:"mtime"`
}

// ChangelogEntry is one CHANGELOG block (heading + body).
type ChangelogEntry struct {
	Heading string `json:"heading"`
	Body    string `json:"body"`
}

// DocEntry mirrors the TS DocEntry interface.
type DocEntry struct {
	ID    string `json:"id"`
	Title string `json:"title"`
	File  string `json:"file"`
	Mtime *int   `json:"mtime,omitempty"` // epoch seconds
}

var (
	decisionLineRe = regexp.MustCompile(`^\s*##\s+(?P<id>\d+)\s*(?:—|-{1,2})\s*(?P<title>.*\S)\s*$`)
	statusRe       = regexp.MustCompile(`(?i)^\s*(?:[-*]\s*)?\**\s*Status\s*\**\s*[:=]\s*(.+\S)\s*$`)
)

// ParseDecisionsMd parses a knowledge/decisions.md into Decision blocks.
func ParseDecisionsMd(text string) []Decision {
	out := []Decision{}
	currentIdx := -1
	body := []string{}

	flush := func() {
		if currentIdx >= 0 {
			out[currentIdx].Body = strings.TrimSpace(strings.Join(body, "\n"))
		}
	}

	for _, line := range strings.Split(text, "\n") {
		if m := decisionLineRe.FindStringSubmatch(line); m != nil {
			flush()
			body = body[:0]
			out = append(out, Decision{ID: m[1], Title: m[2], Status: "", Body: ""})
			currentIdx = len(out) - 1
			continue
		}
		if currentIdx >= 0 {
			if sm := statusRe.FindStringSubmatch(line); sm != nil && out[currentIdx].Status == "" {
				out[currentIdx].Status = strings.TrimSpace(strings.Trim(strings.TrimSpace(sm[1]), "*"))
			}
			body = append(body, line)
		}
	}
	flush()
	return out
}

func mtimeSeconds(path string) int {
	if info, err := os.Stat(path); err == nil {
		return int(info.ModTime().Unix())
	}
	return 0
}

var (
	decisionFileRe = regexp.MustCompile(`^decision-(\d+)`)
	h1Re           = regexp.MustCompile(`^#\s+(.*\S)\s*$`)
	statusLineRe   = regexp.MustCompile(`(?im)^status:\s*(.+\S)\s*$`)
)

func backlogDecisions(repo string) []Decision {
	ddir := filepath.Join(repo, "backlog", "decisions")
	entries, err := os.ReadDir(ddir)
	if err != nil {
		return []Decision{}
	}
	out := []Decision{}
	for _, e := range entries {
		if e.IsDir() || !decisionFileRe.MatchString(e.Name()) {
			continue
		}
		fpath := filepath.Join(ddir, e.Name())
		did := strings.TrimSuffix(e.Name(), ".md")
		if m := decisionFileRe.FindStringSubmatch(e.Name()); m != nil {
			did = m[1]
		}
		text, _ := shared.ReadText(fpath)
		title := strings.TrimSuffix(e.Name(), ".md")
		for _, line := range strings.Split(text, "\n") {
			if hm := h1Re.FindStringSubmatch(line); hm != nil {
				title = strings.TrimSpace(hm[1])
				break
			}
		}
		status := ""
		if sm := statusLineRe.FindStringSubmatch(text); sm != nil {
			status = strings.TrimSpace(sm[1])
		}
		body := text
		if strings.HasPrefix(strings.TrimLeft(text, " \t\r\n"), "---") {
			parts := strings.Split(text, "---")
			if len(parts) >= 3 {
				body = strings.Join(parts[2:], "---")
			}
		}
		mt := mtimeSeconds(fpath)
		out = append(out, Decision{ID: did, Title: title, Status: status, Body: sliceRunes(strings.TrimSpace(body), 8000), Mtime: &mt})
	}
	sortByMtimeDescThenID(out)
	return out
}

// sortByMtimeDescThenID sorts decisions by mtime DESC, then id (localeCompare).
func sortByMtimeDescThenID(d []Decision) {
	sort.SliceStable(d, func(i, j int) bool {
		mi, mj := derefOr0(d[i].Mtime), derefOr0(d[j].Mtime)
		if mi != mj {
			return mi > mj
		}
		return localeLess(d[i].ID, d[j].ID)
	})
}

func derefOr0(p *int) int {
	if p == nil {
		return 0
	}
	return *p
}

func resolveVault() string {
	if env := os.Getenv("OBSIDIAN_VAULT_PATH"); env != "" {
		return env
	}
	if env := os.Getenv("TASKNOTES_VAULT"); env != "" {
		return env
	}
	def := filepath.Join(Home(), "GITHUB", "ObsidianPKM")
	if _, err := os.Stat(def); err == nil {
		return def
	}
	return ""
}

var (
	h23Re     = regexp.MustCompile(`^#{2,3}\s+`)
	hAnyRe    = regexp.MustCompile(`^#\s+`)
	isoDateRe = regexp.MustCompile(`(\d{4}-\d{2}-\d{2})`)
)

// ParseChangelogBlocks parses a CHANGELOG.md into blocks (one block = one
// `##`/`###` heading + body). H1 is the doc title and skipped; empty-body blocks
// dropped. Sorted newest-first (reversed if file appears oldest-first).
func ParseChangelogBlocks(text string, limit int) []ChangelogEntry {
	blocks := []ChangelogEntry{}
	var current *ChangelogEntry
	bodyLines := []string{}

	flush := func() {
		if current != nil {
			body := strings.TrimSpace(strings.Join(bodyLines, "\n"))
			if body != "" {
				current.Body = body
				blocks = append(blocks, *current)
			}
			current = nil
		}
		bodyLines = bodyLines[:0]
	}

	for _, line := range strings.Split(text, "\n") {
		switch {
		case h23Re.MatchString(line):
			flush()
			current = &ChangelogEntry{Heading: strings.TrimSpace(line)}
		case hAnyRe.MatchString(line):
			flush()
		case current != nil:
			bodyLines = append(bodyLines, line)
		}
	}
	flush()

	if len(blocks) >= 2 {
		firstDate := firstSubmatch(isoDateRe, blocks[0].Heading)
		lastDate := firstSubmatch(isoDateRe, blocks[len(blocks)-1].Heading)
		if firstDate != "" && lastDate != "" && firstDate < lastDate {
			for i, j := 0, len(blocks)-1; i < j; i, j = i+1, j-1 {
				blocks[i], blocks[j] = blocks[j], blocks[i]
			}
		}
	}

	if len(blocks) > limit {
		return blocks[:limit]
	}
	return blocks
}

func firstSubmatch(re *regexp.Regexp, s string) string {
	if m := re.FindStringSubmatch(s); m != nil {
		return m[1]
	}
	return ""
}

func vaultChangelog(repo, repoName, vault string, limit int) []ChangelogEntry {
	if localText, ok := shared.ReadText(filepath.Join(repo, "knowledge", "CHANGELOG.md")); ok && localText != "" {
		return ParseChangelogBlocks(localText, limit)
	}
	if vault == "" {
		return []ChangelogEntry{}
	}
	if vaultText, ok := shared.ReadText(filepath.Join(vault, "Efforts", "Work", "dgk", repoName, "CHANGELOG.md")); ok && vaultText != "" {
		return ParseChangelogBlocks(vaultText, limit)
	}
	return []ChangelogEntry{}
}

var (
	fmBlockRe   = regexp.MustCompile(`(?s)^---\n(.*?)\n---`)
	titleLineRe = regexp.MustCompile(`(?im)^title:\s*(.+\S)\s*$`)
	idLineRe    = regexp.MustCompile(`(?im)^id:\s*(.+\S)\s*$`)
	quoteEdgeRe = regexp.MustCompile(`^["']|["']$`)
)

func backlogDocs(repo string) []DocEntry {
	docsDir := filepath.Join(repo, "backlog", "docs")
	out := []DocEntry{}
	entries, err := os.ReadDir(docsDir)
	if err != nil {
		return out
	}
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".md") {
			continue
		}
		fpath := filepath.Join(docsDir, e.Name())
		text, _ := shared.ReadText(fpath)
		base := strings.TrimSuffix(e.Name(), ".md")
		title := base
		var fm string
		if m := fmBlockRe.FindStringSubmatch(text); m != nil {
			fm = m[1]
			if tm := titleLineRe.FindStringSubmatch(fm); tm != nil {
				title = strings.TrimSpace(quoteEdgeRe.ReplaceAllString(tm[1], ""))
			}
		}
		if title == base {
			for _, line := range strings.Split(text, "\n") {
				if hm := h1Re.FindStringSubmatch(line); hm != nil {
					title = strings.TrimSpace(hm[1])
					break
				}
			}
		}
		id := base
		if fm != "" {
			if im := idLineRe.FindStringSubmatch(fm); im != nil {
				id = strings.TrimSpace(im[1])
			}
		}
		mt := mtimeSeconds(fpath)
		out = append(out, DocEntry{ID: id, Title: title, File: e.Name(), Mtime: &mt})
	}
	sort.SliceStable(out, func(i, j int) bool {
		mi, mj := derefOr0(out[i].Mtime), derefOr0(out[j].Mtime)
		if mi != mj {
			return mi > mj
		}
		return localeLess(out[i].ID, out[j].ID)
	})
	return out
}

// CollectKnowledge inspects knowledge/: decisions, lektion-*, memory/*, vault
// CHANGELOG tail, backlog docs. Pass an empty vaultPath to auto-resolve.
func CollectKnowledge(cwd, vaultPath string) map[string]any {
	repo := RepoRoot(cwd)
	kdir := filepath.Join(repo, "knowledge")

	// Decisions: prefer knowledge/decisions.md, else backlog/decisions/*.
	decisions := []Decision{}
	decMdPath := filepath.Join(kdir, "decisions.md")
	if decMdText, ok := shared.ReadText(decMdPath); ok && decMdText != "" {
		decisions = ParseDecisionsMd(decMdText)
		decMdMtime := mtimeSeconds(decMdPath)
		for i := range decisions {
			m := decMdMtime
			decisions[i].Mtime = &m
		}
		sortByMtimeDescThenID(decisions)
	}
	if len(decisions) == 0 {
		decisions = backlogDecisions(repo)
	}

	// lektion-*.md + optional lessons-learned.md — sorted newest-first
	lektionen := []LektionEntry{}
	if entries, err := os.ReadDir(kdir); err == nil {
		for _, e := range entries {
			if !e.IsDir() && lektionFileRe.MatchString(e.Name()) {
				lektionen = append(lektionen, LektionEntry{Name: e.Name(), Mtime: mtimeSeconds(filepath.Join(kdir, e.Name()))})
			}
		}
	}
	lessonsFile := filepath.Join(kdir, "lessons-learned.md")
	if info, err := os.Stat(lessonsFile); err == nil {
		lektionen = append(lektionen, LektionEntry{Name: "lessons-learned.md", Mtime: int(info.ModTime().Unix())})
	}
	sort.SliceStable(lektionen, func(i, j int) bool { return lektionen[i].Mtime > lektionen[j].Mtime })

	// knowledge/memory/*.md — sorted newest-first
	memory := []MemoryEntry{}
	memDir := filepath.Join(kdir, "memory")
	if entries, err := os.ReadDir(memDir); err == nil {
		for _, e := range entries {
			if !e.IsDir() && strings.HasSuffix(e.Name(), ".md") {
				memory = append(memory, MemoryEntry{Name: e.Name(), Mtime: mtimeSeconds(filepath.Join(memDir, e.Name()))})
			}
		}
	}
	sort.SliceStable(memory, func(i, j int) bool { return memory[i].Mtime > memory[j].Mtime })

	// CHANGELOG: prefer knowledge/CHANGELOG.md, fall back to vault path
	vault := vaultPath
	if vault == "" {
		vault = resolveVault()
	}
	repoName := lastPathSegment(repo)
	changelog := vaultChangelog(repo, repoName, vault, 20)

	docs := backlogDocs(repo)

	return map[string]any{
		"available": true,
		"decisions": decisions,
		"lektionen": lektionen,
		"memory":    memory,
		"changelog": changelog,
		"docs":      docs,
	}
}

var lektionFileRe = regexp.MustCompile(`^lektion-.*\.md$`)
